tcms/rpc/tests/test_testplan.py

```diff
@@ -226,3 +228,123 @@ def test_ignores_existing_mappings(self):
+        with self.assertRaisesRegex(XmlRPCFault, 'name'):
+            self.rpc_client.TestPlan.create({
+                'product': product.pk,
+                'product_version': version.pk,
+                'name': '',
+                'type': plan_type.pk,
+                'text': 'Testing TCMS',
+                'parent': None,
+            })
+
+
+class TestCreatePermission(APIPermissionsTestCase):
+    permission_label = 'testplans.add_testplan'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.product = ProductFactory()
+        self.version = VersionFactory(product=self.product)
+        self.plan_type = PlanTypeFactory(name='api testing')
+        self.params = {
+            'product': self.product.pk,
+            'product_version': self.version.pk,
+            'name': 'Testplan foobar',
+            'type': self.plan_type.pk,
+            'text': 'Testing TCMS',
+            'parent': None,
+        }
+
+    def verify_api_with_permission(self):
+        plan = self.rpc_client.TestPlan.create(self.params)
+        self.assertIsNotNone(plan)
+        self.assertEqual(self.params['product'], plan['product_id'])
+        self.assertEqual(self.params['product_version'], plan['product_version_id'])
+        self.assertEqual(self.params['name'], plan['name'])
+        self.assertEqual(self.params['type'], plan['type_id'])
+        self.assertEqual(self.params['text'], plan['text'])
+        self.assertEqual(self.params['parent'], plan['parent'])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.create(self.params)
+
+
+class TestListAttachments(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+    def test_list_attachments_with_wrong_plan_id(self):
+        with self.assertRaisesRegex(XmlRPCFault, 'TestPlan matching query does not exist'):
+            self.rpc_client.TestPlan.list_attachments(-1)
+
+
+class TestListAttachmentsPermissions(APIPermissionsTestCase):
+    permission_label = 'attachments.view_attachment'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def verify_api_with_permission(self):
+        attachments = self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+        self.assertEqual(0, len(attachments))
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegexp(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+
+
+class TestAddAttachment(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def test_add_attachment(self):
+        self.rpc_client.TestPlan.add_attachment(self.plan.pk, 'attachment.txt', 'kiwitcms')
+        attachments = self.rpc_client.TestPlan.list_attachments(self.plan.pk)
```

Comment:
Better use `Attachment.objects.attachments_for_object(self.plan)` here as well. It's a bit faster and makes sure there's only 1 API function that we're concerned with in this scenario.